libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1631 @@
+
+  if (!strv_element_in_json_string_node_list (strv, array_nodes, out_in_array, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+/* This is just a composable way to get the outvalue */
+static gboolean
+invert_outvalue (gboolean *out_value)
+{
+  g_assert (out_value != NULL);
+
+  *out_value = !(*out_value);
+
+  return TRUE;
+}
+
+static GStrv
+eos_updater_override_locales_list (void)
+{
+  const gchar *override_locales = eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_OVERRIDE_LOCALES", NULL);
+
+  if (override_locales)
+    return g_strsplit (override_locales, ";", -1);
+
+  return NULL;
+}
+
+static gboolean
+get_locales_list_from_flatpak_installation (GStrv  *out_strv,
+                                            GError **error)
+{
+  g_autoptr(FlatpakInstallation) installation = eos_updater_get_flatpak_installation (NULL, error);
+  g_assert (out_strv != NULL);
+
+  if (!installation)
+    return FALSE;
+
+  /* TODO: Right now this returns a only the testing override or NULL,
+   * but we might want to do something a little more clever based on what is
+   * supported by Flatpak in future, see
+   * https://github.com/flatpak/flatpak/issues/1156 */
+  *out_strv = eos_updater_override_locales_list ();
+
+  return TRUE;
+}
+
+static gboolean
+action_filter_applies (JsonObject   *object,
+                       const gchar  *filter_key_name,
+                       gboolean     *is_filtered,
+                       GError      **error)
+{
+  const gchar *current_architecture_strv[] =
+  {
+    eos_updater_get_system_architecture_string (),
+    NULL
+  };
+  g_auto(GStrv) supported_languages = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (!get_locales_list_from_flatpak_installation (&supported_languages,
+                                                   error))
+    return FALSE;
+
+  if (g_strcmp0 (filter_key_name, "architectures") == 0 &&
+      strv_element_in_json_member ((GStrv) current_architecture_strv,
+                                   object,
+                                   filter_key_name,
+                                   is_filtered,
+                                   error) &&
+      invert_outvalue (is_filtered))
+    return TRUE;
+
+  if (g_strcmp0 (filter_key_name, "~architectures") == 0 &&
+      strv_element_in_json_member ((GStrv) current_architecture_strv,
+                                   object,
+                                   filter_key_name,
+                                   is_filtered,
+                                   error))
+    return TRUE;
+
+  if (g_strcmp0 (filter_key_name, "locales") == 0 &&
+      strv_element_in_json_member (supported_languages,
+                                   object,
+                                   filter_key_name,
+                                   is_filtered,
+                                   error) &&
+      invert_outvalue (is_filtered))
+    return TRUE;
+
+  if (g_strcmp0 (filter_key_name, "~locales") == 0 &&
```

Comment:
The code here (and in the caller) is not checking for the disallowed condition of having a filter and its inverse. I’ll add that.